// @near-eth/utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildIndexerTxQuery = exports.findNearProof = exports.findEthProof = exports.nearOnEthSyncHeight = exports.ethOnNearSyncHeight = exports.borshifyOutcomeProof = exports.findFinalizationTxOnNear = exports.findFinalizationTxOnEthereum = exports.aurora = exports.nep141 = exports.erc20 = exports.urlParams = void 0;
exports.urlParams = __importStar(require("./url-params"));
exports.erc20 = __importStar(require("./erc20"));
exports.nep141 = __importStar(require("./nep141"));
exports.aurora = __importStar(require("./aurora"));
var findFinalizationTx_1 = require("./findFinalizationTx");
Object.defineProperty(exports, "findFinalizationTxOnEthereum", { enumerable: true, get: function () { return findFinalizationTx_1.findFinalizationTxOnEthereum; } });
Object.defineProperty(exports, "findFinalizationTxOnNear", { enumerable: true, get: function () { return findFinalizationTx_1.findFinalizationTxOnNear; } });
var borshify_proof_1 = require("./borshify-proof");
Object.defineProperty(exports, "borshifyOutcomeProof", { enumerable: true, get: function () { return borshify_proof_1.borshifyOutcomeProof; } });
var ethOnNearClient_1 = require("./ethOnNearClient");
Object.defineProperty(exports, "ethOnNearSyncHeight", { enumerable: true, get: function () { return ethOnNearClient_1.ethOnNearSyncHeight; } });
var nearOnEthClient_1 = require("./nearOnEthClient");
Object.defineProperty(exports, "nearOnEthSyncHeight", { enumerable: true, get: function () { return nearOnEthClient_1.nearOnEthSyncHeight; } });
var findProof_1 = require("./findProof");
Object.defineProperty(exports, "findEthProof", { enumerable: true, get: function () { return findProof_1.findEthProof; } });
Object.defineProperty(exports, "findNearProof", { enumerable: true, get: function () { return findProof_1.findNearProof; } });
var indexer_1 = require("./indexer");
Object.defineProperty(exports, "buildIndexerTxQuery", { enumerable: true, get: function () { return indexer_1.buildIndexerTxQuery; } });
